File: tests/api_loading/consumers/LoadingProbe.swift
import Foundation
import loading_native

typealias ModelLoader = ProbeModelLoader
typealias ModelHandle = ProbeModelHandle
typealias GenerativeModel = ProbeGenerativeModel
typealias ModelFiles = ProbeModelFiles
typealias ModelParts = ProbeModelParts
typealias SamplingDefaults = ProbeSamplingDefaults
typealias GenerationDefaults = ProbeGenerationDefaults
typealias LoadError = ProbeLoadError
typealias Session = ProbeSession
typealias BundleRepo = ProbeBundleRepo
typealias DownloadProgressSink = ProbeDownloadProgressSink

func config(_ backend: String = "cpu", parts: Bool = false) -> LoadConfig {
  LoadConfig(
    contextSize: 24, backend: backend, draftModel: parts ? "missing-probe-draft.gguf" : nil,
    gpuDepthformer: parts)
}

func consumed(_ loader: ModelLoader) {
  do {
    _ = try loader.build()
    fatalError("reused loader succeeded")
  } catch LoadError.Consumed {} catch { fatalError("unexpected reuse error: \(error)") }
  do {
    _ = try loader.buildGenerative()
    fatalError("reused typed loader succeeded")
  } catch LoadError.Consumed {} catch { fatalError("unexpected typed reuse error: \(error)") }
}

func checkInfo(_ model: GenerativeModel, parts: Bool = false) {
  let info = model.info()
  precondition(info.requestedContext == 24 && info.capacity == 24)
  precondition(info.backend == "Cpu" && info.gpuDepthformer == parts)
  precondition(info.draftModel == (parts ? "missing-probe-draft.gguf" : nil))
  if parts {
    precondition(info.chatTemplate == "probe-template")
    precondition(abs(info.temperature - 0.37) < 0.00001 && abs(info.topP - 0.71) < 0.00001)
    precondition(info.topK == 7 && abs(info.minP - 0.13) < 0.00001)
    precondition(abs(info.repetitionPenalty - 1.23) < 0.00001)
  }
}

func runSession(_ session: Session) throws -> [String: Any] {
  try session.append(tokens: [0, 1])
  precondition(session.position() == 2)
  let tokens = try session.generate()
  precondition(tokens.count == 3 && tokens.allSatisfy { $0 < 2 })
  precondition(session.position() == 5)
  return ["tokens": tokens, "position": session.position()]
}

func runDefaults(_ bytes: Data) throws -> [String] {
  let empty = SamplingDefaults(
    temperature: nil, topP: nil, topK: nil, minP: nil, repetitionPenalty: nil)
  let sampling = SamplingDefaults(
    temperature: 0.37, topP: 0.71, topK: 7, minP: 0.13, repetitionPenalty: 1.23)
  var profiles: [(String, GenerationDefaults?, GenerationDefaults)] = [
    ("absent", nil, .text(sampling: empty)),
    ("text-empty", .text(sampling: empty), .text(sampling: empty)),
  ]
  for (name, value) in [
    (
      "audio",
      GenerationDefaults.audio(
        sampling: sampling, numberOfDecodingThreads: 3, audioTemperature: 0.625, audioTopK: 11)
    ),
    (
      "audio",
      .audio(
        sampling: empty, numberOfDecodingThreads: 0, audioTemperature: 0, audioTopK: 0)
    ),
    (
      "audio",
      .audio(
        sampling: empty, numberOfDecodingThreads: UInt32.max, audioTemperature: 1,
        audioTopK: UInt32.max)
    ),
    (
      "audio-empty",
      .audio(
        sampling: empty, numberOfDecodingThreads: nil, audioTemperature: nil, audioTopK: nil)
    ),
  ] {
    profiles.append((name, value, value))
  }
  for (raw, canonical) in [
    (
      " { \"nested\" : [true, null, {\"x\":7}], \"label\": \"line\\ntext\" } ",
      "{\"label\":\"line\\ntext\",\"nested\":[true,null,{\"x\":7}]}"
    ),
    ("[1, 2, null]", "[1,2,null]"), ("42", "42"), ("true", "true"),
    ("\"text\"", "\"text\""), ("null", "null"),
  ] {
    profiles.append(("other", .other(rawJson: raw), .other(rawJson: canonical)))
  }
  for (_, defaults, expected) in profiles {
    for typed in [false, true] {
      let parts = ModelParts(
        model: bytes, multimodalProjector: nil, audioDecoder: nil, audioTokenizer: nil,
        draftModel: nil, inferenceType: nil, chatTemplate: nil, generationDefaults: defaults)
      var loader: ModelLoader? = ModelLoader(source: .parts(parts: parts), config: config())
      var handle: ModelHandle?
      var model: GenerativeModel?
      if typed {
        model = try loader!.buildGenerative()
      } else {
        handle = try loader!.build()
        model = handle!.asGenerative()
      }
      consumed(loader!)
      let observed = model!.generationDefaultsForProbe()
      let session = try model!.createSession()
      weak let releasedLoader = loader
      weak let releasedHandle = handle
      weak let releasedModel = model
      loader = nil
      handle = nil
      model = nil
      precondition(releasedLoader == nil && releasedHandle == nil && releasedModel == nil)
      precondition(observed == expected)
      let result = try runSession(session)
      precondition(result["tokens"] as? [UInt32] == [0, 1, 0])
    }
  }
  for raw in ["", "{", "null trailing", "{\"x\":NaN}"] {
    for typed in [false, true] {
      let parts = ModelParts(
        model: bytes, multimodalProjector: nil, audioDecoder: nil, audioTokenizer: nil,
        draftModel: nil, inferenceType: nil, chatTemplate: nil,
        generationDefaults: .other(rawJson: raw))
      let loader = ModelLoader(source: .parts(parts: parts), config: config())
      do {
        if typed { _ = try loader.buildGenerative() } else { _ = try loader.build() }
        fatalError("malformed defaults succeeded")
      } catch let LoadError.InvalidConfig(field, value, reason, detail) {
        precondition(field == "generation_defaults.raw_json" && value == raw)
        precondition(reason == "invalid_json" && !detail.isEmpty)
      }
      consumed(loader)
    }
  }
  return Set(profiles.map { "parts-defaults-\($0.0)" }).sorted()
    + ["parts-defaults-invalid-json"]
}

func runConfigs(_ bytes: Data) throws -> [String] {
  let profiles: [(String, UInt64?, UInt64, UInt32)] = [
    ("default", nil, 4096, 64), ("zero", 0, 64, 64),
    ("cap", 64, 64, 64), ("cap", 65, 65, 64),
    ("wide", UInt64(UInt32.max) + 25, UInt64(UInt32.max) + 25, 64),
    ("wide", UInt64.max, 64, 64), ("small", 1, 1, 1),
  ]
  for (_, request, observed, capacity) in profiles {
    let options =
      request.map { LoadConfig(contextSize: $0, backend: "cpu") }
      ?? LoadConfig(backend: "cpu")
    precondition(options.contextSize == (request ?? 4096))
    precondition(options.draftModel == nil && !options.gpuDepthformer)
    for typed in [false, true] {
      var loader: ModelLoader? = ModelLoader(source: .bytes(bytes: bytes), config: options)
      var handle: ModelHandle?
      var model: GenerativeModel?
      if typed {
        let typedConfig = EngineConfig(
          contextSize: options.contextSize, backend: .cpu,
          bundleRepo: nil, draftModel: nil, gpuDepthformer: false)
        loader = modelLoaderWithEngineConfig(source: .bytes(bytes: bytes), config: typedConfig)
        model = try loader!.buildGenerative()
      } else {
        handle = try loader!.build()
        model = handle!.asGenerative()
      }
      consumed(loader!)
      let info = model!.info()
      precondition(info.requestedContext == observed && info.capacity == capacity)
      precondition(info.backend == "Cpu" && info.draftModel == nil && !info.gpuDepthformer)
      let session = try model!.createSession()
      weak let releasedLoader = loader
      weak let releasedHandle = handle
      weak let releasedModel = model
      loader = nil
      handle = nil
      model = nil
      precondition(releasedLoader == nil && releasedHandle == nil && releasedModel == nil)
      if capacity == 1 {
        try session.append(tokens: [0])
        do {
          try session.append(tokens: [1])
          fatalError("append exceeded the configured context")
        } catch let LoadError.Engine(detail) { precondition(!detail.isEmpty) }
        precondition(session.position() == 1)
      } else {
        let result = try runSession(session)
        precondition(result["tokens"] as? [UInt32] == [0, 1, 0])
      }
    }
  }
  return [
    "native-config-default", "native-config-zero", "native-config-cap", "native-config-wide",
    "native-config-small",
  ]
}

func runFiles(_ root: URL) throws -> [String] {
  func path(_ name: String) -> String { root.appendingPathComponent(name).path }
  func files(_ primary: String, inference: String?) -> ModelFiles {
    ModelFiles(
      model: path(primary), multimodalProjector: "missing-projector.gguf",
      audioDecoder: path("missing-decoder.gguf"), audioTokenizer: "missing-tokenizer.gguf",
      draftModel: "missing-draft.gguf",
      extras: ["future_file": "future.bin", "absolute_file": path("absolute.bin")],
      inferenceType: inference, chatTemplate: "file-template")
  }
  for typed in [false, true] {
    for inference: String? in [nil, "llama.cpp/text-to-text"] {
      var loader: ModelLoader? = ModelLoader(
        source: .files(files: files("model.gguf", inference: inference)), config: config())
      var handle: ModelHandle?
      var model: GenerativeModel?
      if typed {
        model = try loader!.buildGenerative()
      } else {
        handle = try loader!.build()
        precondition(handle!.kind() == "Generative")
        model = handle!.asGenerative()
      }
      consumed(loader!)
      checkInfo(model!)
      let resolved = model!.files()
      precondition(resolved.model == path("model.gguf"))
      precondition(resolved.multimodalProjector == path("missing-projector.gguf"))
      precondition(resolved.audioDecoder == path("missing-decoder.gguf"))
      precondition(resolved.audioTokenizer == path("missing-tokenizer.gguf"))
      precondition(resolved.draftModel == path("missing-draft.gguf"))
      precondition(
        resolved.extras == [
          "future_file": path("future.bin"), "absolute_file": path("absolute.bin"),
        ])
      precondition(resolved.inferenceType == "llama.cpp/text-to-text")
      precondition(
        resolved.chatTemplate == "file-template" && model!.info().chatTemplate == "file-template")
      let session = try model!.createSession()
      weak let releasedLoader = loader
      weak let releasedHandle = handle
      weak let releasedModel = model
      loader = nil
      handle = nil
      model = nil
      precondition(releasedLoader == nil && releasedHandle == nil && releasedModel == nil)
      let result = try runSession(session)
      precondition(result["tokens"] as? [UInt32] == [0, 1, 0])
    }
  }
  for name in ["kind", "missing", "inference"] {
    for typed in [false, true] {
      let source = files(
        name == "kind" ? "kws.gguf" : "missing-primary.gguf",
        inference: name == "inference" ? "future/unsupported" : nil)
      let loader = ModelLoader(source: .files(files: source), config: config("metal"))
      do {
        if typed { _ = try loader.buildGenerative() } else { _ = try loader.build() }
        fatalError("invalid file load succeeded")
      } catch let LoadError.KindMismatch(expected, actual, architecture) {
        precondition(
          name == "kind" && expected == "Generative" && actual == "Hotword" && architecture == "kws"
        )
      } catch let LoadError.Source(sourceKind, detail) {
        precondition(name == "missing" && sourceKind == "files" && !detail.isEmpty)
      } catch let LoadError.UnsupportedInferenceType(inferenceType) {
        precondition(name == "inference" && inferenceType == "future/unsupported")
      }
      consumed(loader)
    }
  }
  return ["native-files", "native-files-kind", "native-files-missing", "native-files-inference"]
}

@main
struct LoadingProbe {
  static func main() throws {
    let root = URL(fileURLWithPath: CommandLine.arguments[1])
    let bytes = try Data(contentsOf: root.appendingPathComponent("model.gguf"))
    var cases: [String] = []
    var input = Data(count: bytes.count)
    _ = input.withUnsafeMutableBytes { bytes.copyBytes(to: $0) }
    let bufferAddress = input.withUnsafeBytes { UInt(bitPattern: $0.baseAddress!) }
    precondition(bufferAddress != bytes.withUnsafeBytes { UInt(bitPattern: $0.baseAddress!) })
    var loader: ModelLoader? = ModelLoader(source: .bytes(bytes: input), config: config())
    input.resetBytes(in: 0..<input.count)
    precondition(bufferAddress == input.withUnsafeBytes { UInt(bitPattern: $0.baseAddress!) })
    precondition(input.allSatisfy { $0 == 0 })
    var handle: ModelHandle? = try loader!.build()
    consumed(loader!)
    precondition(handle!.kind() == "Generative")
    var first = handle!.asGenerative()
    var second = handle!.asGenerative()
    precondition(first != nil && second != nil)
    checkInfo(second!)
    weak let releasedLoader = loader
    weak let releasedHandle = handle
    weak let releasedFirst = first
    loader = nil
    handle = nil
    first = nil
    precondition(releasedLoader == nil && releasedHandle == nil && releasedFirst == nil)
    let session = try second!.createSession()
    weak let releasedModel = second
    second = nil
    precondition(releasedModel == nil)
    let result = try runSession(session)
    cases.append("bytes-lifetime")

    let defaults = SamplingDefaults(
      temperature: 0.37, topP: 0.71, topK: 7, minP: 0.13, repetitionPenalty: 1.23)
    let parts = ModelParts(
      model: bytes, multimodalProjector: Data([1]), audioDecoder: Data([2]),
      audioTokenizer: Data([3]), draftModel: Data([4]), inferenceType: "llama.cpp/text-to-text",
      chatTemplate: "probe-template", generationDefaults: .text(sampling: defaults))
    let partsLoader = ModelLoader(source: .parts(parts: parts), config: config(parts: true))
    let partsModel = try partsLoader.buildGenerative()
    consumed(partsLoader)
    checkInfo(partsModel, parts: true)
    _ = try runSession(partsModel.createSession())
    cases.append("parts-defaults")
    cases += try runDefaults(bytes)

    let pathLoader = ModelLoader(
      source: .path(path: root.appendingPathComponent("model.gguf").path), config: config())
    let pathModel = try pathLoader.buildGenerative()
    consumed(pathLoader)
    checkInfo(pathModel)
    _ = try runSession(pathModel.createSession())
    cases.append("native-path")
    cases += try runFiles(root)
    cases += try runConfigs(bytes)
    cases += try runProduction(root, bytes: bytes)
    cases += try runRemote(URL(fileURLWithPath: CommandLine.arguments[2]))
    cases += try runProductionRemote(URL(fileURLWithPath: CommandLine.arguments[2]))

    for (arch, kind) in [
      ("bert", "Encoder"), ("modernbert", "Encoder"), ("whisper", "Whisper"), ("silero_vad", "Vad"),
      ("kws", "Hotword"),
    ] {
      for typed in [false, true] {
        let data = try Data(contentsOf: root.appendingPathComponent("\(arch).gguf"))
        let bad = ModelLoader(source: .bytes(bytes: data), config: config("metal"))
        do {
          if typed { _ = try bad.buildGenerative() } else { _ = try bad.build() }
          fatalError("wrong kind succeeded")
        } catch let LoadError.KindMismatch(expected, actual, architecture) {
          precondition(
            expected == "Generative" && actual == kind && architecture == arch,
            "\(arch) kind mismatch payload (typed=\(typed))")
        }
        consumed(bad)
      }
      cases.append("kind-\(arch)")
    }
    for name in ["unknown", "malformed", "backend", "invalid-backend", "assembly", "inference"] {
      for typed in [false, true] {
        let data =
          name == "unknown"
          ? try Data(contentsOf: root.appendingPathComponent("unknown.gguf"))
          : name == "assembly"
            ? try Data(contentsOf: root.appendingPathComponent("llama.gguf"))
            : name == "malformed" ? Data([0, 1]) : bytes
        var unsupportedParts = parts
        unsupportedParts.inferenceType = "future/unsupported"
        let bad = ModelLoader(
          source: name == "inference" ? .parts(parts: unsupportedParts) : .bytes(bytes: data),
          config: config(
            name == "backend" ? "metal" : name == "invalid-backend" ? "invalid-probe" : "cpu"))
        do {
          if typed { _ = try bad.buildGenerative() } else { _ = try bad.build() }
          fatalError("invalid load succeeded")
        } catch let LoadError.UnsupportedArchitecture(architecture) {
          precondition(name == "unknown" && architecture == "future_probe")
        } catch let LoadError.Source(sourceKind, detail) {
          precondition(name == "malformed" && sourceKind == "bytes" && !detail.isEmpty)
        } catch let LoadError.Assembly(backend, detail) {
          precondition(name == "backend" || name == "assembly")
          precondition(backend == (name == "backend" ? "Metal" : "Cpu") && !detail.isEmpty)
        } catch let LoadError.InvalidConfig(field, value, reason, detail) {
          precondition(name == "invalid-backend" && field == "backend")
          precondition(value == "invalid-probe" && reason == "unknown_backend" && !detail.isEmpty)
        } catch let LoadError.UnsupportedInferenceType(inferenceType) {
          precondition(name == "inference" && inferenceType == "future/unsupported")
        }
        consumed(bad)
      }
      cases.append(name)
    }
    let future = futureHandleForProbe()
    precondition(future.kind() == "future-probe" && future.asGenerative() == nil)
    cases.append("future-kind")
    let output = try JSONSerialization.data(
      withJSONObject: ["cases": cases.sorted(), "generation": result], options: [.sortedKeys])
    print(String(decoding: output, as: UTF8.self))
  }
}

File: tests/api_loading/consumers/LoadingProbe.kt
package loadingprobe

import uniffi.cera_ffi.BackendPreference
import uniffi.cera_ffi.EngineConfig
import uniffi.loading_native.LoadConfig
import uniffi.loading_native.Source
import uniffi.loading_native.futureHandleForProbe
import uniffi.loading_native.modelLoaderWithEngineConfig
import java.io.File
import uniffi.loading_native.ProbeGenerationDefaults as GenerationDefaults
import uniffi.loading_native.ProbeGenerativeModel as GenerativeModel
import uniffi.loading_native.ProbeLoadException as LoadException
import uniffi.loading_native.ProbeModelFiles as ModelFiles
import uniffi.loading_native.ProbeModelLoader as ModelLoader
import uniffi.loading_native.ProbeModelParts as ModelParts
import uniffi.loading_native.ProbeSamplingDefaults as SamplingDefaults
import uniffi.loading_native.ProbeSession as Session

private fun config(
    backend: String = "cpu",
    parts: Boolean = false,
) = LoadConfig(24u, backend, if (parts) "missing-probe-draft.gguf" else null, parts)

fun consumed(loader: ModelLoader) {
    for (typed in listOf(false, true)) {
        try {
            if (typed) loader.buildGenerative() else loader.build()
            error("reused loader succeeded")
        } catch (_: LoadException.Consumed) {
            // Both entry points share the consumed state.
        }
    }
}

fun checkInfo(
    model: GenerativeModel,
    parts: Boolean = false,
) {
    val info = model.info()
    check(info.requestedContext == 24uL && info.capacity == 24u)
    check(info.backend == "Cpu" && info.gpuDepthformer == parts)
    check(info.draftModel == if (parts) "missing-probe-draft.gguf" else null)
    if (parts) {
        check(info.chatTemplate == "probe-template")
        check(kotlin.math.abs(info.temperature - 0.37f) < 0.00001f)
        check(kotlin.math.abs(info.topP - 0.71f) < 0.00001f && info.topK == 7u)
        check(kotlin.math.abs(info.minP - 0.13f) < 0.00001f)
        check(kotlin.math.abs(info.repetitionPenalty - 1.23f) < 0.00001f)
    }
}

fun runSession(session: Session): String =
    session.use {
        it.append(listOf(0u, 1u))
        check(it.position() == 2u)
        val tokens = it.generate()
        check(tokens.size == 3 && tokens.all { token -> token < 2u })
        check(it.position() == 5u)
        "{\"tokens\":[${tokens.joinToString(",")}],\"position\":${it.position()}}"
    }

private fun runDefaults(bytes: ByteArray): List<String> {
    val empty = SamplingDefaults(null, null, null, null, null)
    val sampling = SamplingDefaults(0.37f, 0.71f, 7u, 0.13f, 1.23f)
    val profiles =
        mutableListOf<Triple<String, GenerationDefaults?, GenerationDefaults>>(
            Triple("absent", null, GenerationDefaults.Text(empty)),
            Triple("text-empty", GenerationDefaults.Text(empty), GenerationDefaults.Text(empty)),
        )
    for ((name, value) in listOf(
        "audio" to GenerationDefaults.Audio(sampling, 3u, 0.625f, 11u),
        "audio" to GenerationDefaults.Audio(empty, 0u, 0f, 0u),
        "audio" to GenerationDefaults.Audio(empty, UInt.MAX_VALUE, 1f, UInt.MAX_VALUE),
        "audio-empty" to GenerationDefaults.Audio(empty, null, null, null),
    )) {
        profiles += Triple(name, value, value)
    }
    for ((raw, canonical) in listOf(
        " { \"nested\" : [true, null, {\"x\":7}], \"label\": \"line\\ntext\" } " to
            "{\"label\":\"line\\ntext\",\"nested\":[true,null,{\"x\":7}]}",
        "[1, 2, null]" to "[1,2,null]",
        "42" to "42",
        "true" to "true",
        "\"text\"" to "\"text\"",
        "null" to "null",
    )) {
        profiles += Triple("other", GenerationDefaults.Other(raw), GenerationDefaults.Other(canonical))
    }
    for ((_, defaults, expected) in profiles) {
        for (typed in listOf(false, true)) {
            val parts = ModelParts(bytes, null, null, null, null, null, null, defaults)
            val loader = ModelLoader(Source.Parts(parts), config())
            val handle = if (typed) null else loader.build()
            val model = if (typed) loader.buildGenerative() else checkNotNull(handle?.asGenerative())
            consumed(loader)
            val observed = model.generationDefaultsForProbe()
            val session = model.createSession()
            loader.close()
            handle?.close()
            model.close()
            check(observed == expected)
            check(runSession(session) == "{\"tokens\":[0,1,0],\"position\":5}")
        }
    }
    for (raw in listOf("", "{", "null trailing", "{\"x\":NaN}")) {
        for (typed in listOf(false, true)) {
            val parts = ModelParts(bytes, null, null, null, null, null, null, GenerationDefaults.Other(raw))
            ModelLoader(Source.Parts(parts), config()).use { loader ->
                try {
                    if (typed) loader.buildGenerative() else loader.build()
                    error("malformed defaults succeeded")
                } catch (error: LoadException.InvalidConfig) {
                    check(error.field == "generation_defaults.raw_json" && error.value == raw)
                    check(error.reason == "invalid_json" && error.detail.isNotEmpty())
                }
                consumed(loader)
            }
        }
    }
    return profiles.map { "parts-defaults-${it.first}" }.distinct().sorted() + "parts-defaults-invalid-json"
}

private fun runConfigs(bytes: ByteArray): List<String> {
    data class Profile(
        val request: ULong?,
        val observed: ULong,
        val capacity: UInt,
    )
    val profiles =
        listOf(
            Profile(null, 4096uL, 64u),
            Profile(0uL, 64uL, 64u),
            Profile(64uL, 64uL, 64u),
            Profile(65uL, 65uL, 64u),
            Profile(UInt.MAX_VALUE.toULong() + 25uL, UInt.MAX_VALUE.toULong() + 25uL, 64u),
            Profile(ULong.MAX_VALUE, 64uL, 64u),
            Profile(1uL, 1uL, 1u),
        )
    for ((request, observed, capacity) in profiles) {
        val options = if (request == null) LoadConfig(backend = "cpu") else LoadConfig(contextSize = request, backend = "cpu")
        check(options.contextSize == (request ?: 4096uL))
        check(options.draftModel == null && !options.gpuDepthformer)
        for (typed in listOf(false, true)) {
            val loader =
                if (typed) {
                    modelLoaderWithEngineConfig(
                        Source.Bytes(bytes),
                        EngineConfig(options.contextSize, BackendPreference.CPU, null, null, false),
                    )
                } else {
                    ModelLoader(Source.Bytes(bytes), options)
                }
            val handle = if (typed) null else loader.build()
            val model = if (typed) loader.buildGenerative() else checkNotNull(handle!!.asGenerative())
            consumed(loader)
            val info = model.info()
            check(info.requestedContext == observed && info.capacity == capacity)
            check(info.backend == "Cpu" && info.draftModel == null && !info.gpuDepthformer)
            val session = model.createSession()
            loader.close()
            handle?.close()
            model.close()
            if (capacity == 1u) {
                session.use {
                    it.append(listOf(0u))
                    try {
                        it.append(listOf(1u))
                        error("append exceeded the configured context")
                    } catch (error: LoadException.Engine) {
                        check(error.detail.isNotEmpty())
                    }
                    check(it.position() == 1u)
                }
            } else {
                check(runSession(session) == "{\"tokens\":[0,1,0],\"position\":5}")
            }
        }
    }
    return listOf("native-config-default", "native-config-zero", "native-config-cap", "native-config-wide", "native-config-small")
}

private fun runFiles(root: File): List<String> {
    fun path(name: String) = File(root, name).absolutePath

    fun files(
        primary: String,
        inference: String?,
    ) = ModelFiles(
        path(primary),
        "missing-projector.gguf",
        path("missing-decoder.gguf"),
        "missing-tokenizer.gguf",
        "missing-draft.gguf",
        mapOf("future_file" to "future.bin", "absolute_file" to path("absolute.bin")),
        inference,
        "file-template",
    )
    for (typed in listOf(false, true)) {
        for (inference in listOf(null, "llama.cpp/text-to-text")) {
            val loader = ModelLoader(Source.Files(files("model.gguf", inference)), config())
            val handle = if (typed) null else loader.build()
            if (handle != null) check(handle.kind() == "Generative")
            val model = if (typed) loader.buildGenerative() else checkNotNull(handle!!.asGenerative())
            consumed(loader)
            checkInfo(model)
            val resolved = model.files()
            check(resolved.model == path("model.gguf"))
            check(resolved.multimodalProjector == path("missing-projector.gguf"))
            check(resolved.audioDecoder == path("missing-decoder.gguf"))
            check(resolved.audioTokenizer == path("missing-tokenizer.gguf"))
            check(resolved.draftModel == path("missing-draft.gguf"))
            check(resolved.extras == mapOf("future_file" to path("future.bin"), "absolute_file" to path("absolute.bin")))
            check(resolved.inferenceType == "llama.cpp/text-to-text")
            check(resolved.chatTemplate == "file-template" && model.info().chatTemplate == "file-template")
            val session = model.createSession()
            loader.close()
            handle?.close()
            model.close()
            check(runSession(session) == "{\"tokens\":[0,1,0],\"position\":5}")
        }
    }
    for (name in listOf("kind", "missing", "inference")) {
        for (typed in listOf(false, true)) {
            val source =
                files(
                    if (name ==
                        "kind"
                    ) {
                        "kws.gguf"
                    } else {
                        "missing-primary.gguf"
                    },
                    if (name == "inference") "future/unsupported" else null,
                )
            ModelLoader(Source.Files(source), config("metal")).use { loader ->
                try {
                    if (typed) loader.buildGenerative() else loader.build()
                    error("invalid file load succeeded")
                } catch (error: LoadException.KindMismatch) {
                    check(name == "kind" && error.expected == "Generative" && error.actual == "Hotword" && error.architecture == "kws")
                } catch (error: LoadException.Source) {
                    check(name == "missing" && error.sourceKind == "files" && error.detail.isNotEmpty())
                } catch (error: LoadException.UnsupportedInferenceType) {
                    check(name == "inference" && error.inferenceType == "future/unsupported")
                }
                consumed(loader)
            }
        }
    }
    return listOf("native-files", "native-files-kind", "native-files-missing", "native-files-inference")
}

fun main(args: Array<String>) {
    require(args.size == 2) { "Expected model and remote fixture directories" }
    val root = File(args[0])
    val bytes = File(root, "model.gguf").readBytes()
    val cases = mutableListOf<String>()
    val input = bytes.copyOf()
    val loader = ModelLoader(Source.Bytes(input), config())
    input.fill(0)
    val handle = loader.build()
    consumed(loader)
    check(handle.kind() == "Generative")
    val first = checkNotNull(handle.asGenerative())
    val second = checkNotNull(handle.asGenerative())
    checkInfo(second)
    loader.close()
    handle.close()
    first.close()
    val session = second.createSession()
    second.close()
    val result = runSession(session)
    cases += "bytes-lifetime"

    val defaults = SamplingDefaults(0.37f, 0.71f, 7u, 0.13f, 1.23f)
    val parts =
        ModelParts(
            bytes,
            byteArrayOf(1),
            byteArrayOf(2),
            byteArrayOf(3),
            byteArrayOf(4),
            "llama.cpp/text-to-text",
            "probe-template",
            GenerationDefaults.Text(defaults),
        )
    ModelLoader(Source.Parts(parts), config(parts = true)).use { partsLoader ->
        partsLoader.buildGenerative().use { model ->
            consumed(partsLoader)
            checkInfo(model, parts = true)
            runSession(model.createSession())
        }
    }
    cases += "parts-defaults"
    cases += runDefaults(bytes)
    ModelLoader(Source.Path(File(root, "model.gguf").absolutePath), config()).use { pathLoader ->
        pathLoader.buildGenerative().use { model ->
            consumed(pathLoader)
            checkInfo(model)
            runSession(model.createSession())
        }
    }
    cases += "native-path"
    cases += runFiles(root)
    cases += runConfigs(bytes)
    cases += runProduction(root, bytes)
    cases += runRemote(File(args[1]))
    cases += runProductionRemote(File(args[1]))

    for ((arch, kind) in listOf(
        "bert" to "Encoder",
        "modernbert" to "Encoder",
        "whisper" to "Whisper",
        "silero_vad" to "Vad",
        "kws" to "Hotword",
    )) {
        for (typed in listOf(false, true)) {
            ModelLoader(Source.Bytes(File(root, "$arch.gguf").readBytes()), config("metal")).use { bad ->
                try {
                    if (typed) bad.buildGenerative() else bad.build()
                    error("wrong kind succeeded")
                } catch (error: LoadException.KindMismatch) {
                    check(error.expected == "Generative" && error.actual == kind && error.architecture == arch) {
                        "$arch kind mismatch payload (typed=$typed)"
                    }
                }
                consumed(bad)
            }
        }
        cases += "kind-$arch"
    }
    for (name in listOf("unknown", "malformed", "backend", "invalid-backend", "assembly", "inference")) {
        for (typed in listOf(false, true)) {
            val data =
                when (name) {
                    "unknown" -> File(root, "unknown.gguf").readBytes()
                    "assembly" -> File(root, "llama.gguf").readBytes()
                    "malformed" -> byteArrayOf(0, 1)
                    else -> bytes
                }
            val backend =
                when (name) {
                    "backend" -> "metal"
                    "invalid-backend" -> "invalid-probe"
                    else -> "cpu"
                }
            val source = if (name == "inference") Source.Parts(parts.copy(inferenceType = "future/unsupported")) else Source.Bytes(data)
            ModelLoader(source, config(backend)).use { bad ->
                try {
                    if (typed) bad.buildGenerative() else bad.build()
                    error("invalid load succeeded")
                } catch (error: LoadException.UnsupportedArchitecture) {
                    check(name == "unknown" && error.architecture == "future_probe")
                } catch (error: LoadException.Source) {
                    check(name == "malformed" && error.sourceKind == "bytes" && error.detail.isNotEmpty())
                } catch (error: LoadException.Assembly) {
                    check(name == "backend" || name == "assembly")
                    check(error.backend == (if (name == "backend") "Metal" else "Cpu") && error.detail.isNotEmpty())
                } catch (error: LoadException.InvalidConfig) {
                    check(name == "invalid-backend" && error.field == "backend")
                    check(error.value == "invalid-probe" && error.reason == "unknown_backend" && error.detail.isNotEmpty())
                } catch (error: LoadException.UnsupportedInferenceType) {
                    check(name == "inference" && error.inferenceType == "future/unsupported")
                }
                consumed(bad)
            }
        }
        cases += name
    }
    futureHandleForProbe().use { future ->
        check(future.kind() == "future-probe" && future.asGenerative() == null)
    }
    cases += "future-kind"
    println("{\"cases\":[${cases.sorted().joinToString(",") { "\"$it\"" }}],\"generation\":$result}")
}
